### ISSUE TITLE
what changes would we need to do if we used our own dataset?

### ISSUE THREAD

**Issue by awais00012** [explainingai-code/DDPM-Pytorch#1]
Thanks for the awesome explanation. Could you tell me which changes we need before training the model on our data?

**Comment by explainingai-code** [comment]
Hello,

Thanks for the appreciation. I apologize that should have been part of the README , I have updated it now. 
Can you take a look - https://github.com/explainingai-code/DDPM-Pytorch/blob/main/README.md#training-on-your-own-images and let me know in case you face any issues

**Comment by awais00012** [comment]
thanks for the modification in the repo for training the model on custom dataset. however i am facing this issue when i triad  the model on my own data. my dataset with the name of ultrasound256CH1 contain train and test images. all the images have 256*256 size, channel 1.
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/145996791/ec172c80-58c5-4af9-b744-b62fcfc70487)

**Comment by explainingai-code** [comment]
Can you tell me the im_path value you used in the config ?
And also the directory structure of your dataset. Is it $REPO_ROOT/ultrasound256CH1/train/*.png ?

**Comment by awais00012** [comment]
> Can you tell me the im_path value you used in the config ? And also the directory structure of your dataset. Is it $REPO_ROOT/ultrasound256CH1/train/*.png ?
> 
> The error basically means that the code wasn't able to find any png files in the location it was searching.

yes the dataset is in the repo root, still getting this error, kindly how can i solve it.
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/145996791/1b709286-cebe-483b-8342-0daf901966a2)

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/145996791/c22fd3de-ddb4-41b7-ac25-b125fe9eb940)

**Comment by explainingai-code** [comment]
Got it. Create a subfolder 'images' inside train directory and put all training png files in there.
So $REPO_ROOT/ultrasound256CH1/train/images/*.png

Leave the config as it is to point to "ultrasound256CH1/train" .
Can you try that and let me know if it works?

**Comment by awais00012** [comment]
yes i tried, unfortunately it does not work.

**Comment by explainingai-code** [comment]
Can you print the directory and path the code is searching at https://github.com/explainingai-code/DDPM-Pytorch/blob/main/dataset/mnist_dataset.py#L40 and share that.

`print(d_name, os.path.join(im_path, d_name, '*.{}'.format(self.im_ext)))
`
Also comment line https://github.com/explainingai-code/DDPM-Pytorch/blob/main/dataset/mnist_dataset.py#L42

**Comment by awais00012** [comment]
that error has been resolved. that error was occurring  because of the arrangement of the dataset. i created 5 classes of the different images for the train data and used "data/train/" as a path and its worked. 
now encountering this error 
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/145996791/4aa155dc-7a6e-4812-a83b-64e628cc7e80)

**Comment by awais00012** [comment]
that error has been resolved. that error was occurring because of the arrangement of the dataset. i created 5 classes of the different images for the train data and used "data/train/" as a path and its worked.
now encountering this error:
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/145996791/8333d8c7-add2-45da-b857-bd29e84ff6e4)

**Comment by explainingai-code** [comment]
You are training on cpu as of now right ?
Also can you confirm if your conda environment has python3.8 and have the requirements installed as mentioned in https://github.com/explainingai-code/DDPM-Pytorch/tree/main?tab=readme-ov-file#quickstart

**Comment by awais00012** [comment]
hi sir, i kept the batch size 10, just want to run for 40 epochs and the total images are only 828. could you pleases tell me why the model required so heavy computational power(memory) and how can i handle this issue? 

RuntimeError: CUDA out of memory. Tried to allocate 640.00 GiB (GPU 0; 14.75 GiB total capacity; 2.16 GiB already allocated; 11.63 GiB free; 2.17 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF.

**Comment by explainingai-code** [comment]
Its because the images are 256x256 and by default the model config does downsampling only twice.
Few things you can try to ensure you are able to train.

1. Resize the images to 64x64 in the loader and train diffusion model on these 64x64 images
2. Have all three down blocks to downsample by placing down_sample : [True, True, True] in the config
3. Try with  num_mid_layers : 1 in the config

I think that should reduce the model size considerably.

**Comment by awais00012** [comment]
Thanks, the model works very well i trained on my datasets. 
I would request you to add a few more things to the repo for better results and comparative analytical analysis between the original and the generated images like classifier free guidance, exponential moving average, IS and FID score.  looking forward to your outstanding and easy implementation!

**Comment by explainingai-code** [comment]
Yes, I wanted to have this repo as a intro to diffusion which is why didn't want to add those and leave this as a bare minimum diffusion repo. I do plan to create a stable diffusion repo which should have some of these incorporated. Once that is done I will try to put the parts you mentioned here as well (if I am able to do that without adding too much complexity to the current implementation ).

**Comment by thatdev6** [comment]
Hello,

I made all the relevant changes mentioned in the readme and in this thread but after my images are loaded I get an AttributeError. 

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/04b3ee93-339a-4546-8bb7-51dc31715f95)

**Comment by explainingai-code** [comment]
Hello @thatdev6 , this code expects the path to have png files. But seems like thats not the case for the path you have provided. Is it npy file?
Cause In that case you would have to change this [line](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/dataset/mnist_dataset.py#L50)

**Comment by thatdev6** [comment]
> Hello @thatdev6 , this code expects the path to have png files. But seems like thats not the case for the path you have provided. Is it npy file? Cause In that case you would have to change this [line](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/dataset/mnist_dataset.py#L50)

No my path has png files

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/804bbb95-51a6-4833-a930-059707e2c3eb)

**Comment by explainingai-code** [comment]
Are you using the same code or have you made some modifications ? Your list at the end of dataset initialization is a list of numpy.ndarray objects(according to the error), which cannot be because the dataset class during initialization just fetches the filenames.

**Comment by thatdev6** [comment]
Ok so i changed that line and also had to change the im channels to 3, but the model doesnt train. Do you think this issue is related to the filepath issue mentioned above

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/88fe654e-6dcf-4745-ab5d-3ba2859b243f)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/546eb010-fa04-4348-bed6-4f56a12a98cd)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/73e66546-e507-40e6-b794-6b5fb0a2a78e)

**Comment by thatdev6** [comment]
> Are you using the same code or have you made some modifications ? Your list at the end of dataset initialization is a list of numpy.ndarray objects(according to the error), which cannot be because the dataset class during initialization just fetches the filenames.

Yes, I modified the loader function to load and downsample my images. They are rectangular and have a jpg format
I figured out my mistake and it has corrected

This is how modified the loader function
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/b98d0cb8-f79e-4704-bd43-d45669d91b42)


I also changed the im channels to 3, Now I get a runtime error while training
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/97fcb740-a26e-4fc2-b4ba-32f14120df97)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/cfee404f-e02b-46f4-af66-e91580aaf5c5)

**Comment by explainingai-code** [comment]
Before converting to tensor by any chance did you forget to convert the numpy arrays from HxWx3 to 3xWxH ?

**Comment by thatdev6** [comment]
> Before converting to tensor by any chance did you forget to convert the numpy arrays from HxWx3 to 3xWxH ?

How would i do that?

**Comment by thatdev6** [comment]
I also modified the sample function for rectangular images 
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/096a880d-5f85-42e5-a3e0-7d276d02caab)

**Comment by explainingai-code** [comment]
I dont think the 3xwxh is an issue because the error says that your image shapes are 3xWxh so thats fine. But I think your path does not have all same size images. Some images are 3264x2448 and some are 2448x3264 .
Can you check this.

**Comment by thatdev6** [comment]
> I dont think the 3xwxh is an issue because the error says that your image shapes are 3xWxh so thats fine. But I think your path does not have all same size images. Some images are 3264x2448 and some are 2448x3264 . Can you check this.

Yes, I think your right so the solution would be to downsample all of them to 64x64?

**Comment by explainingai-code** [comment]
Yes center square crop of (2448x2448) and then resize to 64x64.
How many images are there in your dataset?

**Comment by thatdev6** [comment]
> Yes center square crop of (2448x2448) and then resize to 64x64. How many images are there in your dataset?

Around 600 images

**Comment by thatdev6** [comment]
These are changes i made to the loader and getitem function, I assume there is no problem here but for some reason the training gets interrupted (^C)

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/3470ed88-da78-40ff-9c44-4c22f488ebcf)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/e61740e0-1b30-485b-9292-497db1fc7536)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/703ebb1e-8fa1-4cb3-b362-062ee3c3b12d)

**Comment by explainingai-code** [comment]
Couple of things. Move the image reading to the data loader get_item method just like the code in repo. Simply collect the filenames from their. You can do the cropping and resize also in get_item method.
Secondly can you check why its printing "Found 19 images" when actually it should be 600.

**Comment by thatdev6** [comment]
> Couple of things. Move the image reading to the data loader get_item method just like the code in repo. Simply collect the filenames from load_images method and nothing else. You can do the cropping and resize also in get_item method. Secondly can you check why its printing "Found 19 images" when actually it should be 600.

Okay so first of all i should leave the loader function as it is just modify for the jpg images, secondly i should do the image formatting in the get item function
It says found 19 images because at the moment i have only uploaded a subset of the dataset, It was quite annoying to wait for the images to load only to encounter an error in training

**Comment by thatdev6** [comment]
I guess i have encountered the final boss
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/9715c0cf-2a70-4bdc-808b-317f02603116)

I have sampled down my images to 64x64 and also have the changes you mentioned to the previous person before me but still  am running out of memory
Any suggestion?
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/07ce723a-b875-400c-9a92-1f4de460009a)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/9faf9016-c881-4faa-b623-314d1ca4fd49)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/499288e7-0c46-4078-9560-d0947ff9e391)

**Comment by explainingai-code** [comment]
Reduce the batch size to 16. That should work I think.

**Comment by thatdev6** [comment]
> Reduce the batch size to 16. That should work I think.

torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 4.00 GiB. GPU 0 has a total capacity of 14.75 GiB of which 2.31 GiB is free. Process 85462 has 12.44 GiB memory in use. Of the allocated memory 9.43 GiB is allocated by PyTorch, and 2.87 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

**Comment by thatdev6** [comment]
> Reduce the batch size to 16. That should work I think.

It started to train on batch size 4

**Comment by thatdev6** [comment]
I cannot generate the results
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/1cc42a1e-0743-4501-b57f-480976557832)

**Comment by thatdev6** [comment]
I have also tried changing my dataset, the new data set had square images of size 128x128. Yet i encountered the same runtime error while generating the pictures

**Comment by explainingai-code** [comment]
Whats the shape of xt that you get [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/sample_ddpm.py#L20) and noisy_im [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/train_ddpm.py#L72) ?

**Comment by thatdev6** [comment]
> Whats the shape of xt that you get [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/sample_ddpm.py#L20) and noisy_im [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/train_ddpm.py#L72) ?

I was able to solve this one out. I had to reshape the tensors. But I do have a question will the reshaping affect my results?

**Comment by thatdev6** [comment]
after training on the complete dataset this is my output(x0)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/35e9cd4f-b909-465a-bc41-22ee6281cb94)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/854c0297-ec59-45b2-a793-361b003dd562)

**Comment by explainingai-code** [comment]
I dont think you would need to do any reshaping. Can you just let me know the shapes of those two tensors? They should be of the same shape.
These results are after how many epochs of training ?

**Comment by thatdev6** [comment]
> I dont think you would need to do any reshaping. Can you just let me know the shapes of those two tensors? They should be of the same shape. These results are after how many epochs of training ?

After 40 epochs of training, okay I am posting the shapes

**Comment by thatdev6** [comment]
> Whats the shape of xt that you get [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/sample_ddpm.py#L20) and noisy_im [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/train_ddpm.py#L72) ?

Shape of noisy_im: torch.Size([4, 3, 64, 64])
Shape of xt: torch.Size([100, 3, 28, 28])

**Comment by explainingai-code** [comment]
So if you set xt as 100x3x64x64, you are saying you get an error?

**Comment by thatdev6** [comment]
> So if you set xt as 100x3x64x64, you are saying you get an error?

I didnt explicitly set it this from the unmodified code where i get the error when i start generating

**Comment by explainingai-code** [comment]
This was being it as 100x3x28x28 because the config has im_size=28 and the code was for training on mnist(28x28).
It will be the same shape as your noisy images during training, so for your case xt should be number of samplesx3x64x64

**Comment by thatdev6** [comment]
> This was being it as 100x3x28x28 because the config has im_size=28 and the code was for training on mnist(28x28). It will be the same shape as your noisy images during training, so for your case xt should be number of samplesx3x64x64

okay that makes sense but I dont understand it my sample size is different han the configured sample size so do i have to change the configuration as at the momment i am generating 1oo samples after  my reshaping but i should be generating 4. did i get that right?

also does the reshaping affect my results?

**Comment by explainingai-code** [comment]
You can generate how many samples you wish, you would not need any reshaping anywhere.
Assuming you have trained your diffusion model on images of size 3x64x64.
Then xt during sampling should be Nx3x64x64 where N is the number of samples you want. 

Can you clarify where/why you require reshaping? Maybe I am misunderstanding something.

**Comment by thatdev6** [comment]
> I cannot generate the results ![image](https://private-user-images.githubusercontent.com/128231244/313420131-1cc42a1e-0743-4501-b57f-480976557832.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTAwNTgsIm5iZiI6MTcxMDY4OTc1OCwicGF0aCI6Ii8xMjgyMzEyNDQvMzEzNDIwMTMxLTFjYzQyYTFlLTA3NDMtNDUwMS1iNTdmLTQ4MDk3NjU1NzgzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxN1QxNTM1NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZmZlYzFiMzhjNTRhYjhlY2RjNzNjMzJhN2E2OTczNzlhZDlmZDkxYTk4ZTQ1MTZiOTQ4MmU4ZGMxMjg4M2EwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Xn07k4672X3VT_J4pcyVxwalyHYbE06KbqRsnHKlaFU)

after i had trained my model and got to generation the code showed me this output, after debugging i saw that the sizes in the unet file were indeed different so i made the following chanages
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/32d36011-039e-4f07-8f5f-92cade53f3bd)

after this i got another error leading me to the linear scheduler file and there were to some size mismatches so i made the following changes
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/7f5e07a3-f65f-4c87-9282-572cd588b0fc)

after i this i could generate the images and they full of noise, so at the moment i am training on BW images to see if i get better results

**Comment by explainingai-code** [comment]
Okay, I would suggest to remove all the resize portions that you added in both these places, the could should work without any of these. If its not then their might be some other issue but no resize is required anywhere other than dataset class.
After you have made changes in data loader to resize the images(assuming its WxH )
You should train the model on 3xWxH and sample with xt being Nx3xWxH

**Comment by thatdev6** [comment]
The BW images generated the same result this is at x0
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/4a9cd7fb-3fe3-402b-8c98-a059133bbb8d)

**Comment by thatdev6** [comment]
> Okay, I would suggest to remove all the resize portions that you added in both these places, the could should work without any of these. If its not then their might be some other issue but no resize is required anywhere other than dataset class. After you have made changes in data loader to resize the images(assuming its WxH ) You should train the model on 3xWxH and sample with xt being Nx3xWxH

Okay, so firstly remove all the modifications in the unet and linear scheduler, but then how would i resolve the errors while generating?

**Comment by explainingai-code** [comment]
For images of different sizes, you would continue to do what you were already doing in your dataset class. That is centre crop and resize to 64x64.
Then train your model on these 64x64 images and generate an image with xt as 64x64.
To avoid any confusion you can also change the im_size [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/config/default.yaml#L11) to 64(instead of 28).

So basically the only change from the repo code that you should need to do is that centre cropping.

**Comment by thatdev6** [comment]
> For images of different sizes, you would continue to do what you were already doing in your dataset class. That is centre crop and resize to 64x64. Then train your model on these 64x64 images and generate an image with xt as 64x64. To avoid any confusion you can also change the im_size [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/config/default.yaml#L11) to 64(instead of 28).
> 
> So basically the only change from the repo code that you should need to do is that centre cropping. You dont need to actually train a model to test this, just run the sampling script on a randomly initialized model and see if it still throws an error. If it does just share it here and I will take a look.

Okay i will get back to you with that, also should i change the im channels to 3 for rgb pictures?

**Comment by thatdev6** [comment]
> For images of different sizes, you would continue to do what you were already doing in your dataset class. That is centre crop and resize to 64x64. Then train your model on these 64x64 images and generate an image with xt as 64x64. To avoid any confusion you can also change the im_size [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/config/default.yaml#L11) to 64(instead of 28).
> 
> So basically the only change from the repo code that you should need to do is that centre cropping. You dont need to actually train a model to test this, just run the sampling script on a randomly initialized model and see if it still throws an error. If it does just share it here and I will take a look.

Okay so I was able to get outputs now without changing anything except the changing im size and im channels in config and data loader function.

But do you know the reason i am getting such noisy outputs? . Currently batch size is 4 and epochs is 50
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/ab37d317-0f5a-4aec-b8bc-fd15a1bcd10b)

**Comment by explainingai-code** [comment]
This is actually fine. When I was also training on an rgb dataset , I first used to get these outputs, so more training should lead to actual outputs(close to dataset images) ultimately.
In these generation results, when you see x0_400/500 were they closer to your actual dataset images?
Also did you end up training for more epochs? Did the outputs improve?

**Comment by thatdev6** [comment]
When you say continous training, how would I achieve that, doesn't the model overwrite itself, I am actually using the free resources of Google CoLab for training so how could I save my progress and train multiple times for better results.

**Comment by explainingai-code** [comment]
I mean train it for more epochs like 200 epochs.

**Comment by thatdev6** [comment]
> I mean train it for more epochs like 200 epochs.

Okay so multiple epochs, also do you think the reduced batch size is effecting results?

**Comment by explainingai-code** [comment]
Thats something you would have to experiment with and tune but I am assuming you are anyway limited by compute and 4 is the max you can go ?

**Comment by thatdev6** [comment]
Yes 4 is the highest I can go

**Comment by thatdev6** [comment]
> Yes 4 is the highest I can go, I will obviously need help in tuning the model parameters for better results so I will keep you updated.
> 
> Thank you so much for the help in the mean time.

Is there any way i can save my progress while training, what i want to do is say train up to 130 epochs stop my training and then continue training from 130 epochs again?

**Comment by thatdev6** [comment]
this is my result after 100 epochs on batch 6 
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/b23a47fb-371b-402a-917e-3cafd344ac75)

Why are the generated outputs getting so dark, shouldnt they start to mimic the actual images at some point

> This is actually fine. When I was also training on an rgb dataset , I first used to get these outputs, so more training should lead to actual outputs(close to dataset images) ultimately. In these generation results, when you see x0_400/500 were they closer to your actual dataset images? Also did you end up training for more epochs? Did the outputs improve?

**Comment by thatdev6** [comment]
This is at x0_500
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/83191e6f-c58d-424a-8f4b-4894441972d5)

And this is at x0_999
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/5fbe778a-04f4-4e05-8ef3-e3ecec65112a)

**Comment by thatdev6** [comment]
for reference these are some dataset images
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/e5a92564-1468-4591-a023-b212e5791325)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/68529b2b-98dd-4b81-a480-a75ca16596e4)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/2cf2e223-7184-46ce-9ea5-64bfa3c031fc)

**Comment by thatdev6** [comment]
I would also like to mention that as of now my training time per epoch is 40s so to train 100 epochs it took me a little over an hour, I remember in your yt video you trained your model for around 3 hours on 60 epochs, so my dataset might also be a limiting factor over here(257 images after split)
Let me know what think of this conclusion, also how would you update the model params to get better results

**Comment by explainingai-code** [comment]
->Is there any way i can save my progress while training, what i want to do is say train up to 130 epochs stop my training and then continue training from 130 epochs again?
   For resuming training, this should be already happening when the code is loading the checkpoint [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/train_ddpm.py#L49) So all you would need to do is after 130 epochs, download the checkpoint and before running the next 130 epochs, simply place this downloaded checkpoint in the right path.
->so my dataset might also be a limiting factor over here(257 images after split).
Yes 250 images are very less, the number that I gave was for training mnist, 60000 images of 28x28 on NvidiaV100, where I trained for ~50 epochs with batch size 64, so which means about 50000 steps. With 250 images and batch size 6, effectively that means 40 steps in one epoch so even after 100 epochs thats actually just 4000 steps. 
I would suggest doing the following things:
1. If you can get more data then that would definitely help.
2. Continue training for longer if by analyzing the x0-x999 images you are continuing to see improvement in models generation capabilities. 
3. Use augmentations, flipping , and even cropping, based on your images, I am assuming your goal is to generate images with those cracks, so you could take multiple crops from the same images like below

<img width="536" alt="Screenshot 2024-03-19 at 8 51 38 AM" src="https://github.com/explainingai-code/DDPM-Pytorch/assets/144267687/dfa12621-9648-43ed-aad4-de501f169bc2">

**Comment by thatdev6** [comment]
So 

> ->Is there any way i can save my progress while training, what i want to do is say train up to 130 epochs stop my training and then continue training from 130 epochs again? For resuming training, this should be already happening when the code is loading the checkpoint [here](https://github.com/explainingai-code/DDPM-Pytorch/blob/main/tools/train_ddpm.py#L49) So all you would need to do is after 130 epochs, download the checkpoint and before running the next 130 epochs, simply place this downloaded checkpoint in the right path. 

Are the checkpoints saved in the .pth file? So I would have to transfer that file before starting training then

**Comment by explainingai-code** [comment]
Yes download the .pth file after one iteration of training, and put it back in the necessary path before starting the second iteration.

**Comment by thatdev6** [comment]
Okay, I work on the suggestions, in your opinion what is the better alternative 
Opting for a larger dataset with images from different sources and sizes 
Or a the same images cropped to form a larger dataset

**Comment by explainingai-code** [comment]
I would say a larger dataset is beneficial and then using cropping you can further increase the images that diffusion model gets to see during training.

**Comment by thatdev6** [comment]
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/fadca403-682e-489d-bcc5-9f97bc0f6b0a)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/7a8a2418-e83a-4ecc-b1b9-7af3f9a65985)
![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/4ef841da-91b8-4f73-9b0b-772ecb8281bf)

these are my images after crop and resize in the data loader function are these okay?, i think they have lost a lot of quality

Also i do i have a larger dataset around 45k images but the images in it are to inconsistent in every manner size, quality
so what effect will my my current crop in data loader function will have on images which are not 3264 x 2448 and Size: 2448 x 3264

**Comment by explainingai-code** [comment]
For the 45K images, my guess is that the centre crop and resize to 64x64 should handle the inconsistencies in size(I dont know how inconsistent they are in quality). But if you have 45K iamges, I would say why not try with that just to see how best of a output quality you get from ddpm.

**Comment by thatdev6** [comment]
Yes that is the goal right, But training those images will take a considerate amount of time and if the output is still noisy, all the time spent will have been wasted

> For the 45K images, my guess is that the centre crop and resize to 64x64 should handle the inconsistencies in size(I dont know how inconsistent they are in quality). But if you have 45K iamges, I would say why not try with that just to see how good of an output quality you get from ddpm.

**Comment by explainingai-code** [comment]
Yes diffusion models require decent amount of data to train so unless you throw in more compute power, I dont see any other way to reduce the time.  
The single color outputs are normal to occur during the training process(as I mentioned it happened during my training as well). And I would suggest to not compare based on epochs, 100 epochs on a dataset of 200 is not the same as 100 epochs on a dataset of 50000 images. Rather use the number of steps/iterations that has happened in your training, in your case its just 5K steps(compared to the 80K steps that I used in the video). I dont think you need as many steps as that cause you have lesser variation in your dataset but still just wanted to give a perspective.

**Comment by thatdev6** [comment]
> Yes diffusion models require decent amount of data to train so unless you throw in more compute power, I dont see any other way to reduce the time. The single color outputs are normal to occur during the training process(as I mentioned it happened during my training as well). And I would suggest to not compare based on epochs, 100 epochs on a dataset of 200 is not the same as 100 epochs on a dataset of 50000 images. Rather use the number of steps/iterations that has happened in your training, in your case its just 5K steps(compared to the 80K steps that I used in the video). I dont think you need as many steps as that cause you have lesser variation in your dataset but still just wanted to give a perspective.

![image](https://github.com/explainingai-code/DDPM-Pytorch/assets/128231244/e650f310-ff03-44ad-b7ac-d9b97fae0b52)
I am finally starting to see results on the 257 image dataset after 400 epochs of training. I am assuming there are always going to be noisy images in the generated batch even if i train on the 45k images for around 500 epochs.

Also is there a way I can rescale the generated images into a higher resolution after generation.

**Comment by explainingai-code** [comment]
Great. I dont think that assumption is correct, once your model has converged(looks like that point maybe somewhere around 1000 epochs) it will not have these noisy images at all. And with 45K images around after 200 epochs you will most likely be able to see decent outputs(obv you will have to experiment to assert that).

For higher resolution, you can try with pre-trained super resolution models(just searching on the web should give you some models and you can test them to see how good of a result you are getting with that.
Or you can go the stable diffusion route by first training autoencoder to convert from 256x256 to 32x32, training diffusion models on 32x32 images, generating latent 32x32 images at sampling and then decoding them to 256x256 using autoencoder. But that would require training multiple(diffusion and autoencoder) models.

**Comment by thatdev6** [comment]
Hello its me again
I have a question, can this model generate images based on prompts, for example say I prompt to generate an image with snow and cracks, is this achievable?

**Comment by dangdinh17** [comment]
hello, i have a question with my model
i  want to train on my data with the config params like that 
![image](https://github.com/user-attachments/assets/6c145c25-c591-449a-bdc2-54c6ab0797f4)
and my image is in jpg format. I have tried another methods like Dataparallel but it doesn't work.
so please help me with this

**Comment by explainingai-code** [comment]
Yeah, can you try with 64x64 . In config set im_size as 64 and in dataset class's getitem method, resize the image to 64x64. 
Can you see if that gets rid of the error ?

**Comment by dangdinh17** [comment]
i have tried with 64x64 and it worked. but i want to train with the shape 128x128 because of my study. let me introduce about my study and can you give me some essential suggests, please?
i have a dataset with the image in 150x150 image, and i have a blured images with the motion blur and i want to use DDPM model to restore the image the the highest quality. so i want to train my model with the shape 128x128 or 256x256 to suit my data shape. so i have some questions that:
+ if i devided my image (for example 128x128) into 64x64 parts and train with this 64x64 part, so when i used this trained model with the shape 128x128, does it work good?
+ i've tried many model like GANs but the results didn't good, so now i try with the DDPM and may be the next stable diffusion, can you give me some must try models for my study?
thank you

**Comment by explainingai-code** [comment]
If your images are from a dataset(or match a dataset) that has a super resolution model available, then you can train DDPM on 64x64 images and then use that super resolution model checkpoint to get 128x128 image. Or you can train a super resolution model yourself.
The second option is like you said to try with LDM, so then your trained auto-encoder will take care of converting 256x256 images to 64x64 latents(as well as converting 64x64 generated latents to 256x256). And diffusion model will need to be trained on the smaller, 64x64 latent images.

**Comment by dangdinh17** [comment]
oh yes, i see, thank you so much

**Comment by dangdinh17** [comment]
i have another question that, if my data is about motion blur or exposure blur, so if i only use the original code for this blur, does it  work? or i must train the model with adding more noise like motion blur and light blur rather than only gaussian noise?

**Comment by explainingai-code** [comment]
@dangdinh17 , apologies for the late reply. But I have responded to your issue on Stable Diffusion repo. Do take a look at the repo mentioned in that reply - https://github.com/explainingai-code/StableDiffusion-PyTorch/issues/21#issuecomment-2248046225 . I think that implementation does exactly what you need.

**Comment by xXCoffeeColaXc** [comment]
@dangdinh17 Hey, just dont allow the network to compute attentions on resolution 128. Only in lower resolution. This should solve your memory issues.

**Comment by T02097** [comment]
Hi, I would like to ask where do I need to make changes if my dataset is a color image? For example the MIRFlickR dataset

**Comment by explainingai-code** [comment]
Hello @JiangHaiting , 
Changing below config lines with `im_channels=3` and correct `im_size` should be all you need.
https://github.com/explainingai-code/DDPM-Pytorch/blob/b524321c101af13f28a5d1f4fbcca1a847c19924/config/default.yaml#L10-L11

Additionally if your images are larger size and you are training ddpm on resized images or cropped images, then you would have to do resize/crop in dataset class as well.
https://github.com/explainingai-code/DDPM-Pytorch/blob/b524321c101af13f28a5d1f4fbcca1a847c19924/dataset/mnist_dataset.py#L50-L51

**Comment by T02097** [comment]
Ok, thank you very much for your answer, I will give it a try!
&nbsp;


宀
***@***.***







&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2025年1月18日(星期六) 中午12:45
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [explainingai-code/DDPM-Pytorch] what changes would we need to do if we used our own dataset? (Issue #1)















你好@江海婷,
用以下代码更改配置行下面的内容im_channels=3并更正im_size应该是你所需要的一切。
https://github.com/explainingai-code/DDPM-Pytorch/blob/b524321c101af13f28a5d1f4fbcca1a847c19924/config/default.yaml#L10-L11

此外，如果您的图像尺寸较大，并且您正在对ddpm进行调整大小或裁剪图像的培训，那么您也必须在数据集课程中进行调整大小/裁剪。
https://github.com/explainingai-code/DDPM-Pytorch/blob/b524321c101af13f28a5d1f4fbcca1a847c19924/dataset/mnist_dataset.py#L51

-
直接回复此邮件，在GitHub上查看它，或取消订阅.
***@***.*** &gt;
